### PR TITLE
[Bugfix] Resolve softmax and sdp overflow for large values

### PR DESCRIPTION
**Pull request by silvenachen**
## Description ##
The PR solves the issue related to numerical overflow during softmax calculation and the scaled_dot_product_attention that utilizes it. In addition, a similar overflow error occurs in the numpy implementation used in `allo/tests/test_nn.py` that generate golden results for testing. The PR fixes this issue by adding a correct `np_softmax()` function in the testing script.

### Problems ###
- **Softmax Overflow**: The softmax calculation could result in overflow when handling large input values, causing NaNs in hte output.
- **Incorrect Numpy Softmax Implementation**: In the original test script, `np_out = np.exp(inp) / np.exp(inp).sum(axis=1, keepdims=True)` will cause similar overflow problems in the exponentiation computation.

### Proposed Solutions ###
The softmax function has been updated to subtract the maximum value in each row before exponentiation to cope with larger values correctly.

### Examples ###
```
(allo) sl3558@brg-zhang-xcel:/work/shared/users/ugrad/sl3558/mdlm/dit_numpy_allo$ pytest allo/tests/test_nn.py::test_softmax
========================================================= test session starts =========================================================
platform linux -- Python 3.12.4, pytest-8.2.2, pluggy-1.5.0
rootdir: /work/shared/users/ugrad/sl3558/mdlm/dit_numpy_allo/allo
collected 1 item                                                                                                                      

allo/tests/test_nn.py .                                                                                                         [100%]

========================================================== 1 passed in 2.91s ==========================================================
```

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [IR], [Builder], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage (It would be better to provide ~2 different test cases to test the robustness of your code)
- [ ] Code is well-documented

### CODE REVIEW

File: allo/library/nn.py [hunk 1]
```diff
@@ -6,17 +6,24 @@
...
-    Z: Ty[D, D]
-    E: Ty[D, D]
-    S: Ty[D] = 0.0
-
-    for i, j in dsl.grid(D, D, name="exp_sum"):
-        E[i, j] = dsl.exp(X[i, j])
+def softmax[Ty, L](X: "Ty[L, L]") -> "Ty[L, L]":
+    Z: Ty[L, L]
+    E: Ty[L, L]
+    M: Ty[L] = -10000.0
```
Comment (chhzh123) [inline allo/library/nn.py]: Maybe use a larger constant number, say `-0x3f3f3f3f.0` (a large enough number smaller within the range of int32)

File: tests/test_nn.py [hunk 2]
```diff
@@ -51,14 +51,20 @@ def test_linear_float():
...
+    e_x = np.exp(x - x_max)
+    return e_x / np.sum(e_x, axis=axis, keepdims=True)
+
+
 def test_softmax():
     from allo.library.nn import softmax
 
     s = allo.customize(softmax, instantiate=[float32, 8])
     mod = s.build()
     inp = np.random.randn(8, 8).astype(np.float32)
```
Comment (chhzh123) [inline tests/test_nn.py:65]: Can you also use a larger test case? I remember you used some large random tensors in your own test.

**Review by chhzh123 — COMMENTED** [review]
Thanks for contributing to Allo! Just some minor issues

File: allo/library/nn.py (hunk 1 above)
Comment (silvenachen) [reply allo/library/nn.py]: I could not directly use -0x3f3f3f3f.0, but I used another value which I believe is also negative enough,

File: tests/test_nn.py (hunk 2 above)
Comment (silvenachen) [reply tests/test_nn.py:65]: Sure! I added `inp = 1000 * inp` to create larger tensor test cases.

**Review by chhzh123 — APPROVED** [review]
LGTM. Thanks! This PR fixes #176.

**Pull request merged**